### PR TITLE
Handle horovod errors

### PR DESCRIPTION
**Pull request by yuxihu**
This PR populates error originated in Horovod back to MXNet engine through callback. The error will be thrown in MXNet when a sync action is performed, such as wait_to_read(). This approach has no performance impact. This is dependent on a [PR](https://github.com/apache/incubator-mxnet/pull/13922) in MXNet. 

The approaches using [synchronize](https://github.com/ctcyang/horovod/pull/19) and C++ [future](https://github.com/ctcyang/horovod/pull/22) library shows 80~85% performance regression.

### CODE REVIEW

**Review by apeforest — APPROVED** [review]
Left two small comments. Otherwise LGTM

File: horovod/mxnet/adapter.h [hunk 1]
```diff
@@ -16,6 +16,8 @@
 #ifndef HOROVOD_MXNET_ADAPTER_H
 #define HOROVOD_MXNET_ADAPTER_H
 
+#include <mxnet/base.h>
+
```
Comment (apeforest) [inline horovod/mxnet/adapter.h]: nit: remove blank line

Comment (yuxihu) [reply horovod/mxnet/adapter.h]: do we need to have a blank line between third party library and header files from itself?

File: horovod/mxnet/mpi_ops.cc [hunk 2]
```diff
@@ -13,21 +13,21 @@
...
 #include "cuda_util.h"
 #include "mpi_ops.h"
 #include "ready_event.h"
+#include "status_manager.h"
 #include "tensor_util.h"
 
 namespace horovod {
 namespace mxnet {
 
+static StatusManager status_manager;
```
Comment (apeforest) [inline horovod/mxnet/mpi_ops.cc]: Do want want to move the op_count into the status_manager and make it a tuple with the last_error like `(op_count, last_error)`. Would it be helpful for debugging?

Comment (yuxihu) [reply horovod/mxnet/mpi_ops.cc]: This is a good one. Maybe we can save the name (which contains the op_count) to make it more informative.

File: horovod/mxnet/adapter.h (hunk 1 above)
Comment (apeforest) [reply horovod/mxnet/adapter.h]: fine.

**Review by ctcyang — APPROVED** [review]
LGTM. Does this mean MXNet profiler will have data for communication calls too?

**Comment by yuxihu** [comment]
@ctcyang As we discussed offline, MX_API_BEGIN/MX_API_END was used to catch C++ errors and make it possible to throw MXNetError at Python level. It is not profiler related as far as I know.

**Comment by alsrgv** [comment]
TBH, it seems a bit hacky.  This doesn't resolve the situation when the last Horovod operation fails.

I have another idea.  I see that [this code](https://github.com/apache/incubator-mxnet/blob/4649bfa641ad4129b3e83ea0af14b489e512f8f4/src/kvstore/kvstore_local.h#L406) can push operation in the engine, which may throw `dmlc::Error` from `UniqueImpl`.

How would that error be handled?  What if our Horovod error handling will enqueue another high-priority operation which will just throw `dmlc::Error`?

**Comment by alsrgv** [comment]
New API is way nicer! :-)

**Comment by yuxihu** [comment]
@alsrgv We will trigger the Horovod CI once the dependent [PR](https://github.com/apache/incubator-mxnet/pull/13922) is merged.

File: horovod/mxnet/mpi_ops.cc [hunk 3]
```diff
@@ -119,17 +117,18 @@ void DoAllgatherCudaOnCPU(NDArray* tensor, NDArray* output, std::string& name,
   auto enqueue_result = EnqueueTensorAllgather(
       hvd_context, hvd_cpu_tensor, ready_event,
       name, CPU_DEVICE_ID,
-      [hvd_cpu_output, output, cb](const Status& status) {
+      [hvd_cpu_output, output, on_complete](const Status& status) {
         TensorUtil::CopyCPUToCuda(hvd_cpu_output->tensor(), output);
-        cb();
+          on_complete(status.ok() ? nullptr : status.reason().c_str());
```
Comment (apeforest) [inline horovod/mxnet/mpi_ops.cc]: nit: spacing does not align with line above

Comment (yuxihu) [reply horovod/mxnet/mpi_ops.cc]: Done

File: examples/mxnet_imagenet_resnet50.py [hunk 4]
```diff
@@ -67,24 +66,22 @@
...
-parser.add_argument('--optimizer', type=str, default='nag',
-                    help='optimizer to use for training (default: nag)')
 parser.add_argument('--eval-epoch', action='store_true', default=False,
                     help='evaluate validation accuracy after each epoch (default: False)')
 parser.add_argument('--no-cuda', action='store_true', default=False,
                     help='disables CUDA training (default: False)')
 parser.add_argument('--log-interval', type=int, default=0,
                     help='number of batches to wait before logging (default: 0)')
-parser.add_argument('--save-frequency', type=int, default=0,
+parser.add_argument('--save-frequency', type=int, default=10,
```
Comment (apeforest) [inline examples/mxnet_imagenet_resnet50.py]: description says default: 0?

File: examples/mxnet_imagenet_resnet50.py [hunk 5]
```diff
@@ -303,7 +300,7 @@ def main():
                         'lr_scheduler': lr_sched}
     if args.dtype == 'float16':
         optimizer_params['multi_precision'] = True
-    opt = mx.optimizer.create(args.optimizer, sym=out, **optimizer_params)
+    opt = mx.optimizer.create('sgd', sym=out, **optimizer_params)
```
Comment (apeforest) [inline examples/mxnet_imagenet_resnet50.py]: why not let user pick optimizing algorithm?

File: examples/mxnet_imagenet_resnet50.py (hunk 4 above)
Comment (yuxihu) [reply examples/mxnet_imagenet_resnet50.py]: change to 10 to save the checkpoint by default

File: examples/mxnet_imagenet_resnet50.py (hunk 5 above)
Comment (yuxihu) [reply examples/mxnet_imagenet_resnet50.py]: each optimizer has its own optimizer_params. very hard to support all of them

File: horovod/mxnet/mpi_ops.cc [hunk 6]
```diff
@@ -54,55 +60,56 @@ void DoAllreduce(NDArray* tensor, NDArray* output, const std::string& name,
...
+
   // Make async copy of input tensor to CPU tensor and record completion event.
   auto hvd_cpu_buffer = std::make_shared<MXTemporaryBuffer<NDArray>>(
       CPU_DEVICE_ID, tensor->dtype());
   TensorUtil::AsyncCopyCudaToCPU(tensor, hvd_cpu_buffer->tensor());
-  auto ready_event = std::make_shared<MXReadyEvent<NDArray>>(tensor);
-
   auto hvd_context = std::make_shared<MXOpContext<NDArray>>(
       CPU_DEVICE_ID, hvd_cpu_buffer->tensor());
+  auto ready_event = std::make_shared<MXReadyEvent<NDArray>>(tensor);
```
Comment (apeforest) [inline horovod/mxnet/mpi_ops.cc]: Does the order or this line make difference?

Comment (yuxihu) [reply horovod/mxnet/mpi_ops.cc]: nope. just cosmetic change and followed the order in another function.

**Pull request merged**